File: execution/convert_excel_to_json.js
const fs = require('fs');
const path = require('path');
const XLSX = require('xlsx');

const SOURCE_DIR = 'amazon_economics';
const OUTPUT_FILE = 'temp_ingest/weekly_data.jsonl';

function normalizeColumnName(name) {
    if (!name) return "";
    let clean = name.toString().replace(/[^a-zA-Z0-9]/g, '_');
    clean = clean.replace(/_+/g, '_').replace(/^_|_$/g, '');
    return clean.toLowerCase();
}

function extractDatesFromFilename(filename) {
    const match = filename.match(/(\d{1,2}-\d{1,2}-\d{2}) to (\d{1,2}-\d{1,2}-\d{2})/);
    if (match) {
        return { start: match[1], end: match[2] };
    }
    return { start: null, end: null };
}

function formatDate(dateStr) {
    // Input M-D-YY or MM-DD-YY, Output YYYY-MM-DD
    if (!dateStr) return null;
    const parts = dateStr.split('-');
    const m = parts[0].padStart(2, '0');
    const d = parts[1].padStart(2, '0');
    let y = parts[2];
    if (y.length === 2) y = '20' + y;
    return `${y}-${m}-${d}`;
}

function processFile(filePath) {
    console.log(`Processing ${filePath}...`);
    const filename = path.basename(filePath);
    const dates = extractDatesFromFilename(filename);
    
    if (!dates.start) {
        console.warn(`Skipping ${filename}: No dates found in filename.`);
        return [];
    }

    const workbook = XLSX.readFile(filePath);
    const sheetName = workbook.SheetNames[0];
    const sheet = workbook.Sheets[sheetName];
    
    // Convert to JSON with array of arrays to handle headers manually
    const data = XLSX.utils.sheet_to_json(sheet, { header: 1 });
    
    if (data.length < 2) return [];

    const row0 = data[0]; // Category
    const row1 = data[1]; // Metric
    
    const headers = [];
    let lastCat = "info";

    for (let i = 0; i < row0.length; i++) {
        let cat = row0[i];
        let metric = row1[i];
        
        if (cat && !cat.toString().startsWith('Unnamed')) {
            lastCat = cat;
        }
        
        let colName;
        if (!metric || metric.toString().startsWith('Unnamed')) {
            colName = lastCat;
        } else {
            colName = `${lastCat}_${metric}`;
        }
        
        headers.push(normalizeColumnName(colName));
    }

    // Process rows starting from index 2
    const records = [];
    for (let i = 2; i < data.length; i++) {
        const row = data[i];
        // Skip empty rows
        if (!row || row.length === 0) continue;

        const record = {};
        let isEmpty = true;
        
        headers.forEach((header, index) => {
            let val = row[index];
            if (val !== undefined && val !== null && val !== "") {
                isEmpty = false;
            }
            record[header] = val;
        });

        if (!isEmpty) {
            record['report_start_date'] = formatDate(dates.start);
            record['report_end_date'] = formatDate(dates.end);
            record['source_file'] = filename;
            records.push(record);
        }
    }
    
    return records;
}

function main() {
    if (!fs.existsSync('temp_ingest')) {
        fs.mkdirSync('temp_ingest');
    }
    
    const files = fs.readdirSync(SOURCE_DIR).filter(f => f.endsWith('.xlsx'));
    let allData = [];
    
    files.forEach(f => {
        const filePath = path.join(SOURCE_DIR, f);
        try {
            const records = processFile(filePath);
            allData = allData.concat(records);
        } catch (e) {
            console.error(`Error processing ${f}:`, e);
        }
    });
    
    const outStream = fs.createWriteStream(OUTPUT_FILE);
    allData.forEach(entry => {
        outStream.write(JSON.stringify(entry) + '\n');
    });
    outStream.end();
    
    console.log(`Written ${allData.length} rows to ${OUTPUT_FILE}`);
}

main();

File: execution/convert_excel_to_json.py
import os
import re
import glob
import datetime
import json
import pandas as pd
import traceback

SOURCE_DIR = 'amazon_economics'
OUTPUT_FILE = 'temp_ingest/weekly_data.jsonl'

def normalize_column_name(name):
    if not name or pd.isna(name):
        return ""
    clean = re.sub(r'[^a-zA-Z0-9]', '_', str(name))
    clean = re.sub(r'_+', '_', clean).strip('_')
    return clean.lower()

def extract_dates_from_filename(filename):
    basename = os.path.basename(filename)
    match = re.search(r'(\d{1,2}-\d{1,2}-\d{2}) to (\d{1,2}-\d{1,2}-\d{2})', basename)
    if match:
        start_str, end_str = match.groups()
        return start_str, end_str
    return None, None

def process_file(file_path):
    print(f"Processing {file_path}...")
    start_date, end_date = extract_dates_from_filename(file_path)
    if not start_date:
        return []

    try:
        # Read with multi-level header
        df = pd.read_excel(file_path, header=[0, 1], engine='openpyxl')
        
        # Flatten headers
        new_cols = []
        last_cat = "info"
        for col in df.columns:
            cat = col[0]
            metric = col[1]
            
            if not str(cat).startswith('Unnamed'):
                last_cat = str(cat)
            
            if str(metric).startswith('Unnamed'):
                col_name = last_cat
            else:
                col_name = f"{last_cat}_{metric}"
            
            new_cols.append(normalize_column_name(col_name))
            
        df.columns = new_cols
        
        # Metadata
        df['report_start_date'] = pd.to_datetime(start_date, format='%m-%d-%y').strftime('%Y-%m-%d')
        df['report_end_date'] = pd.to_datetime(end_date, format='%m-%d-%y').strftime('%Y-%m-%d')
        df['source_file'] = os.path.basename(file_path)
        
        # Convert to list of dicts
        # Replace NaN with None. astype(object) is needed to allow None in numeric columns
        return df.astype(object).where(pd.notnull(df), None).to_dict(orient='records')

    except Exception as e:
        print(f"Error {file_path}:")
        traceback.print_exc()
        return []

def main():
    if not os.path.exists('temp_ingest'):
        os.makedirs('temp_ingest')
        
    files = glob.glob(os.path.join(SOURCE_DIR, "*.xlsx"))
    all_data = []
    
    for f in files:
        file_data = process_file(f)
        all_data.extend(file_data)
        
    with open(OUTPUT_FILE, 'w') as f:
        for entry in all_data:
            # Handle datetime objects that might still be in the dict if any
            json.dump(entry, f, default=str)
            f.write('\n')
            
    print(f"Written {len(all_data)} rows to {OUTPUT_FILE}")

if __name__ == "__main__":
    main()
